glean/src/server_events.rs

```diff
@@ -0,0 +1,309 @@
+    // metadata for event in `extra_keys`
+}
+
+// Implementing the EventsPingEvent trait for the generated struct SyncstorageGetCollectionsEvent
+impl EventsPingEvent for SyncstorageGetCollectionsEvent {
+    /// Create a GleanEvent for the above-defined Event struct (SyncstorageGetCollectionsEvent).
+    /// Any metadata `extra` values are passed into the extra HashMap.
+    fn glean_event(&self) -> GleanEvent {
+        // Any `extra_keys` will be output below to be inserted into `extra`.
+        // If there are none, an empty, immutable HashMap is created.
+        let extra: HashMap<String, String> = HashMap::new();
+
+        new_glean_event("syncstorage", "get_collections", extra)
+    }
+}
+
+/// Marker trait for events per ping
+pub trait EventsPingEvent {
+    fn glean_event(&self) -> GleanEvent; // Returns an instance of GleanEvent
+}
+
+pub struct EventsPing {
+    pub syncstorage_device_family: String, // Device family from which sync action was initiated. Desktop PC, Tablet, Mobile, and Other.
+    pub syncstorage_hashed_device_id: String, // Hashed device id that is associated with a given account.
+    pub syncstorage_hashed_fxa_uid: String, // User identifier. Uses `hashed_fxa_uid` for accurate count of sync actions.
+    pub syncstorage_platform: String, // Platform from which sync action was initiated. Firefox Desktop, Fenix, or Firefox iOS.
+    pub event: Option<Box<dyn EventsPingEvent>>, // valid event of `EventsPingEvent` for this ping.
+}
+
+// Record and submit `events` ping
+impl GleanEventsLogger {
+    /// General `record_events_ping` function
+    pub fn record_events_ping(&self, request_info: &RequestInfo, params: &EventsPing) {
+        // Define the outer `Metrics` map that holds the metric type.
+        let mut metrics = Metrics::new();
+        // Create the inner metric value map to insert into `Metrics`.
+        // Create corresponding metric value maps to insert into `Metrics`.
+        let mut string_map: HashMap<String, serde_json::Value> = std::collections::HashMap::new();
+        string_map.insert(
+            "syncstorage.device_family".to_string(),
+            serde_json::Value::String(params.syncstorage_device_family.to_string()),
+        );
+        string_map.insert(
+            "syncstorage.hashed_device_id".to_string(),
+            serde_json::Value::String(params.syncstorage_hashed_device_id.to_string()),
+        );
+        string_map.insert(
+            "syncstorage.hashed_fxa_uid".to_string(),
+            serde_json::Value::String(params.syncstorage_hashed_fxa_uid.to_string()),
+        );
+        string_map.insert(
+            "syncstorage.platform".to_string(),
+            serde_json::Value::String(params.syncstorage_platform.to_string()),
+        );
+        metrics.insert("string".to_string(), string_map);
+
+        let mut events: Vec<GleanEvent> = Vec::new();
+        if let Some(event) = &params.event {
+            events.push(event.glean_event());
```

Comment:
Unless we have a lot of different `EventsPingEvent` types, this will always return basically the same thing, a default GleanEvent. (Even then, there's a good argument that YAGNI and you could use a default pattern for it as well.)

Could we just do 
```rust
   events.push(GleanEvent{
    category: "syncstorage".to_owned(),
    name: "get_collections".to_owned(),
    ..Default::default()
   })
```
here?